x-pack/auditbeat/module/system/socket/state.go

```diff
@@ -521,13 +495,21 @@ func (s *state) logStateLoop() {
 	}
 }
 
-func (s *state) ExpireOlder() {
+func (s *state) ExpireFlows() {
+	start := s.clock()
+	toReport := s.expireFlows()
+	if sent := s.reportFlows(&toReport); sent > 0 {
+		s.log.Debugf("ExpireOlder took %v reported=%d", s.clock().Sub(start), sent)
+	}
+}
+
+func (s *state) expireFlows() (toReport linkedList) {
```

Comment:
I'm wondering at this point, when it is now a matter of collecting items, if it wouldn't be simpler to have a `[]linkedElement` for `reportFlows` to work on; the allocations would likely be reduced (particularly if the size hints were used to preallocate) and the code would be simpler. This would involve changes in `onSockTerminated` and `onFlowTerminated` as well.

Reply:
I'm not sure I follow. Removing elements from linkedLists and creating a new linked list from those existing elements is a zero-alloc operation. I'll add a test to ensure this is always the case.

Reply:
I think checking the escape analysis would be worthwhile. The head that is being obtained for adding to `toReport` is not the same memory that is in the source since we are not using the pointer, so it must be allocated. Since it is being used after the function returns, it cannot be on the stack, so I think the allocation needs to be on the heap. The tail of that head will all be already allocated and so they are certainly zero-alloc.

Reply:
You are right; I've looked at the escape analysis and it doesn't escape.